var/spack/repos/builtin/packages/legion/package.py

```diff
@@ -19,79 +19,317 @@ class Legion(CMakePackage):
-    variant('shared', default=True, description='Build shared libraries')
-    variant('hdf5', default=True, description='Enable HDF5 support')
+    version('cr', branch='control_replication')
+
+    depends_on("cmake@3.16:", type='build')
+    # TODO: Need to spec version of MPI v3 for use of the low-level MPI transport
+    # layer. At present the MPI layer is still experimental and we discourge its
+    # use for general (not legion development) use cases.
+    depends_on('mpi', when='network=mpi')
+    depends_on('mpi', when='network=gasnet')  # MPI is required to build gasnet (needs mpicc).
+    depends_on('ucx', when='conduit=ucx')
+    depends_on('mpi', when='conduit=mpi')
+    depends_on('cuda@10.0:11.9', when='+cuda_unsupported_compiler')
+    depends_on('cuda@10.0:11.9', when='+cuda')
+    depends_on('hdf5', when='+hdf5')
+    depends_on('hwloc', when='+hwloc')
+
+    # cuda-centric
+    # reminder for arch numbers to names: 60=pascal, 70=volta, 75=turing, 80=ampere
+    # TODO: we could use a map here to clean up and use naming vs. numbers.
+    cuda_arch_list = ('60', '70', '75', '80')
+    for nvarch in cuda_arch_list:
+        depends_on('kokkos@3.3.01+cuda+cuda_lambda+wrapper cuda_arch={0}'.format(nvarch),
+                   when='%gcc+kokkos+cuda cuda_arch={0}'.format(nvarch))
+        depends_on("kokkos@3.3.01+cuda+cuda_lambda~wrapper cuda_arch={0}".format(nvarch),
+                   when="%clang+kokkos+cuda cuda_arch={0}".format(nvarch))
+
+    depends_on('kokkos@3.3.01~cuda', when='+kokkos~cuda')
+    depends_on("kokkos@3.3.01~cuda+openmp", when='kokkos+openmp')
+
+    depends_on('python@3', when='+python')
+    depends_on('papi', when='+papi')
+    depends_on('zlib', when='+zlib')
+
+    # TODO: Need a AMD/HIP variant to match support landing in 21.03.0.
+
+    # Network transport layer: the underlying data transport API should be used for
+    # distributed data movement.  For Legion, gasnet is the currently the most
+    # mature.  We have many users that default to using no network layer for
+    # day-to-day development thus we default to 'none'.  MPI support is new and
+    # should be considered as a beta release.
+    variant('network', default='none',
+            values=('gasnet', 'mpi', 'none'),
+            description="The network communications/transport layer to use.",
+            multi=False)
+
+    # We default to automatically embedding a gasnet build. To override this
+    # point the package a pre-installed version of GASNet-Ex via the gasnet_root
+    # variant.
+    #
+    # make sure we have a valid directory provided for gasnet_root...
+    def validate_gasnet_root(value):
+        if value == 'none':
+            return True
+
+        if not os.path.isdir(value):
+            print("gasnet_root:", value, "-- no such directory.")
+            return False
+        else:
+            return True
+
+    variant('gasnet_root',
```

Comment:
> Would you be able to use the `gasnet` package already provided in Spack? If users prefer a non-Spack version, then typically we'd recommend adding an external entry to `packages.yaml`.

After discussing this with the gasnet team (as we were updating the legion package), we agreed to no longer use the gasnet package.  It doesn't meet some of our particular needs and we switched our approach to use the preferred configuration and installation path recommended (supported) by the gasnet developers.

Reply:
> we switched our approach to use the preferred configuration and installation path recommended (supported) by the gasnet developers.

To confirm: it looks like the approach is to use a vendored `gasnet` (e.g. using `-DLegion_EMBED_GASNet=ON` and `gasnet_root=none` by default)?

In general we prefer to avoid this since multiple packages depending on `gasnet` (e.g. `tau` and `legion`) might run into problems if they use differing instances of `gasnet`. This appears to be supported by the maintainers of the Spack `legion` and `gasnet` packages though, so I'm inclined to approve it in this case.

Are there future plans to reintroduce the Spack `gasnet` package? Or is there a link to the discussion that led you to decide to remove it?

Reply:
>  is there a link to the discussion that led you to decide to remove it?

Hi @scheibelp : Legion is now defaulting to embedding their own copy of GASNet via CMake, the build of which is customized in non-trivial ways to match Legion's needs and (for the shared library case) embedded directly into the Realm shared library. This `gasnet_root` variant is a "backdoor" in the Legion package that allows a power user to tweak the GASNet build (e.g. applying source patches), but IIUC this is expected to be used rarely/never.

Speaking as a primary maintainer of GASNet, Pat's approach is the Right Way to use GASNet (as a tightly coupled embedded library). GASNet has a highly multi-dimensional and client-dependent configure space, doesn't support same-process "mixing" of clients who select different points in that configure space, and doesn't officially support shared libraries. For these and other reasons, GASNet really doesn't "fit" well as a Spack package. 

The gasnet Spack package was created without consulting the GASNet developers and against our wishes, however my name appears as the package maintainer-of-record. Speaking in that role, the gasnet Spack package is a mess and should be reworked or preferably retired. We do not endorse Spack's gasnet package in any way and would rather it not be used by anyone. We are happy that our only major consumer (Legion) is dropping their reliance on the gasnet spackage, and intend to pursue the rework or removal of Spack's gasnet package as a followup to this PR.